expr/func_aspercent.go

```diff
@@ -0,0 +1,241 @@
+				if _, ok := totalSeries[key]; !ok {
+					serie1.QueryPatt = fmt.Sprintf("asPercent(%s,MISSING)", serie1.QueryPatt)
+					serie1.Target = fmt.Sprintf("asPercent(%s,MISSING)", serie1.Target)
+					serie1.Tags = map[string]string{"name": serie1.Target}
+					for i := range serie1.Datapoints {
+						serie1.Datapoints[i].Val = math.NaN()
+					}
+				} else {
+					serie2 := totalSeries[key]
+					serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+					serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+					serie1.Tags = map[string]string{"name": serie1.Target}
+					for i := range serie1.Datapoints {
+						serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+					}
+				}
+				outSeries = append(outSeries, serie1)
+				cache[Req{}] = append(cache[Req{}], serie1)
+			}
+		}
+		return outSeries, nil
+	}
+
+	var totalsSerie models.Series
+	if math.IsNaN(s.totalFloat) && totals == nil {
+		totalsSerie = sumSeries(series)
+		if len(series) == 1 {
+			totalsSerie.Target = fmt.Sprintf("sumSeries(%s)", totalsSerie.Target)
+			totalsSerie.QueryPatt = fmt.Sprintf("sumSeries(%s)", totalsSerie.QueryPatt)
+			totalsSerie.Tags = map[string]string{"name": totalsSerie.Target}
+		}
+	} else if totals != nil {
+		if len(totals) == 1 {
+			totalsSerie = totals[0]
+		} else if len(totals) == len(series) {
+			sort.Slice(series, func(i, j int) bool {
+				return series[i].Target < series[j].Target
+			})
+			sort.Slice(totals, func(i, j int) bool {
+				return totals[i].Target < totals[j].Target
+			})
+			for i, serie1 := range series {
+				serie2 := totals[i]
+				serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+				serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+				serie1.Tags = map[string]string{"name": serie1.Target}
+				for i := range serie1.Datapoints {
+					serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+				}
+				outSeries = append(outSeries, serie1)
+				cache[Req{}] = append(cache[Req{}], serie1)
+			}
+			return outSeries, nil
+		} else {
+			return nil, errors.New("asPercent second argument must be missing, a single digit, reference exactly 1 series or reference the same number of series as the first argument")
+		}
+	} else {
+		totalsSerie.QueryPatt = fmt.Sprint(s.totalFloat)
+	}
+
+	for _, serie := range series {
+		serie.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie.QueryPatt, totalsSerie.QueryPatt)
+		serie.Target = fmt.Sprintf("asPercent(%s,%s)", serie.Target, totalsSerie.QueryPatt)
+		serie.Tags = map[string]string{"name": serie.Target}
+		for i := range serie.Datapoints {
+			var totalVal float64
+			if len(totalsSerie.Datapoints) > i {
+				totalVal = totalsSerie.Datapoints[i].Val
+			} else {
+				totalVal = s.totalFloat
+			}
+			serie.Datapoints[i].Val = computeAsPercent(serie.Datapoints[i].Val, totalVal)
```

Comment:
I think this violates https://github.com/grafana/metrictank/blob/master/expr/NOTES#L35

Reply:
fixed all and added a test